Allow execution of custom scripts to install packages

iilyak opened the pull request:
This change adds an ability to install additional packages useful
for development. This is done via placing additional scripts into
`bin/extra/` directory. The scripts are named based on package
manager used in a given distribution:
 - `apt-*` for apt based distributions
 - `pkg-*` for FreeBSD
 - `yum-*` - for Centos

The scripts are executed in alphabetical order.

# Testing recommendations

1. create script
    ```
    mkdir -p bin/extra
    cat << EOF > bin/extra/apt-rg.sh
    curl -LO https://github.com/BurntSushi/ripgrep/releases/download/0.8.1/ripgrep_0.8.1_amd64.deb
    sudo dpkg -i ripgrep_0.8.1_amd64.deb
    EOF
    chmod +x bin/extra/apt-rg.sh
    ```
2. Build platform `./build.sh platform debian-stretch`
3. Check the logs you should see following:
    ```
    Preparing to unpack ripgrep_0.8.1_amd64.deb ...
    Unpacking ripgrep (0.8.1) ...
    Setting up ripgrep (0.8.1) ...
    ```
4. Run container 
    ```
    docker run -it couchdbdev/debian-stretch-erlang-19.3.6 bash
    ```
5. Check that `rg` is available
    ``` 
    which rg
    /usr/local/bin/rg
    ```

wohali commented:
@iilyak i am going to finish the 2.2.0 release then come back to review this and the other PR, hope that's OK.

Thanks so much for putting this together!

May I suggest `brew-*` for OSX?

iilyak commented:
> May I suggest brew-* for OSX?

We don't have support for MacOS at the moment 
https://github.com/cloudant/couchdb-ci/blob/master/bin/install-dependencies.sh#L101

Do you want to me to just call `run_scripts ${EXTRA_SCRIPTS_DIR} 'brew-'`? Then we would have to remove exit 1.

I think it would be better to add it when we would have (which is out of the scope for this PR) 
- bin/brew-erlang.sh
- bin/brew-dependencies.sh

Just in case I would decide to work on implementing `brew-dependencies.sh` are you planing to support build of rpm and deb on mac?

wohali commented on bin/install-dependencies.sh:
```diff
@@ -55,6 +58,17 @@ if [[ ${EUID} -ne 0 ]]; then
   exit 1
 fi
 
+# Unfortunatelly run-parts is not available in centos 6.
```
:( sad.

Misspelling of `Unfortunately`. ;)

wohali commented on bin/install-dependencies.sh:
```diff
@@ -55,6 +58,17 @@ if [[ ${EUID} -ne 0 ]]; then
+    local prefix=$2
+    scripts=$(LC_ALL=C; echo ${scripts_dir%/}/${prefix}*[^~,] | xargs -n 1 | sort -n)
+    for script in $scripts ; do
+        [[ -f ${script} && -x ${script} ]] && ${script} || true
```
This will force-succeed if the script fails. That seems wrong? Wouldn't it be better to drop the `|| true` portion?

wohali commented on bin/install-dependencies.sh:
```diff
@@ -55,6 +58,17 @@ if [[ ${EUID} -ne 0 ]]; then
+function run_scripts() {
+    local scripts_dir=$1
+    local prefix=$2
+    scripts=$(LC_ALL=C; echo ${scripts_dir%/}/${prefix}*[^~,] | xargs -n 1 | sort -n)
```
So the idea is you could create:
```
pkg-10-install-erlang.sh
pkg-20-install-elixir-dependent-on-erlang.sh
```

 etc? Nice.

wohali commented:
BTW rebase with master to fix Travis.

wohali commented:
@iilyak If you want to put in the effort to build the Ubuntu/Debian debs and CentOS rpms on Mac, then fine, but as I don't have one, I can't sign up to support it.

iilyak commented:
> @iilyak  If you want to put in the effort to build the Ubuntu/Debian debs and  CentOS rpms on Mac, then fine, but as I don't have one, I can't sign up  to support it.

I more interested in use of `bin/install-dependencies.sh` standalone to install development environment for all new CouchDB contributors (if they find it useful). Currently it is quite hard to setup development environment.

wohali commented:
Sounds great, looking forward to us having support across a broad spectrum of UNIX-like platforms with `bin/install-dependencies.sh`.

iilyak commented:
I changed the `run_scripts` a little bit to make sure it causes whole script to fail (per your request) and fixed a problem when extra scripts directory doesn't exists.
